async-context/README.md

```diff
@@ -0,0 +1,375 @@
+===================================
+|    ...host code...      |
+---------------------------
+|    ...host code...      |
+===================================
+|   Host Continuation B()         |
+===================================
+```
+
+Some notes about the pictures above:
+  - We've introduced a "Root Continuation" as the bottom frame.  This  illustrates a basic assumption that all code is executing in the context of a `Continuation Invocation`.
+  - Multiple `continuations` can be on the stack at the same time, which result in mulitple `Continuation Invocations`.
+
+Here, we've updated our pictures of runtime callstacks, giving  labels to the `Continuation Invocation` instances:
+
+```
+---------------------------                    -----------                            
+|   setTimeout()          |                              |
+---------------------------                              |
+|   function f1()         |                              |
+---------------------------                              \/
+|    ...host code...      |                  Continuation Invocation ci1
+---------------------------                              /\  
+|    ...host code...      |                              |
+===================================                      |
+|    Host Continuation A()        |                      |
+===================================            -----------
+```
+
+and
+
+```
+---------------------------                    -----------
+| function x()            |                              |
+---------------------------                              \/                   
+| function myLoggingAPI() |                  Continuation Invocation ci3                        
+===================================                      /\
+|    Continuation f2()            |                      |
+===================================            -----------
+|    ...host code...      |                              |
+---------------------------                              \/
+|    ...host code...      |                   Continuation Invocation ci2                   
+===================================                      /\
+|   Host Continuation B()         |                      |
+===================================            -----------
+```
+
+## Continuation Invocation States
+A `Continuation Invocation` can be in a number of states:
+  - `pending` - A  `Continuation Invocation` instance has been created, but is not yet executing, nor is it ready to execute.
+  - `cancelled` - A `Continuation Invocation` was abandoned.
+  - `ready` - A `Continuation Invocation` is ready to execute, but is not yet executing
+  - `executing` - `Continuation Invocation` is currently executing on the stack
+  - `paused` - A `Continuation Invocation` has started execution, but is not currently on the stack.  E.g., for generator functions. // todo - is this necessary for async/await & generators.  what is tc39 terminology?
```

Comment:
@Flarna - I think so.  I'll review implementation in Chakra & the spec, but I believe Generator functions yield & restore the call stack, so by definitions above, the generator invocation will be in a single `continuation invocation`.  Again, I need to work through more examples with generators & async/await, so if you have an example where this doesn't make sense, please let me know.

Reply:
I'm also not sure here. Thinking of  `await` I would wonder that a single `Continuation Invocation` for following async function is correct:
```
async function foo() {
  // A
  await some();
  // B
  await thing();
  // C
  await else();
  // D
}
```
Here I would expect that each parts A, B, C and D have their own invocation and that they are chained.

Reply:
> Here I would expect that each parts A, B, C and D have their own invocation and that they are chained

@Flarna - This is my intuition as well.  See comment [here](https://github.com/nodejs/diagnostics/pull/197#issuecomment-394458286).  Certainly the case if you look at above example as simple promises instead of async/await, e.g.,: 

```javascript
async function foo() {
  // A
  return some().then(() => {
      // B
      return thing();
  }).then(()=> {
     // C
     return else();
  }).then(() => {
     // D   
  })
}
```

I'm going to dig into the async/await details later this week & I'll update then if any of this changes.